Add beolink_join source_id parameter to Bang & Olufsen

mj23000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add new "source_id" parameter to the existing `beolink_join` custom action. 

This allows Mozart devices in Home Assistant to join a specific source on other Beolink compatible devices, such as our previous ASE platform and the Beolink Converter NL/ML.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36174

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mj23000 committed:
Add source as parameter to beolink join service

mj23000 committed:
Add beolink join source and responses

mj23000 committed:
Improve comment
Add translation

mj23000 committed:
Merge branch 'dev' into bang_olufsen_beolink_join_source

mj23000 committed:
Remove result from beolink join custom action

mj23000 committed:
Cleanup

mj23000 committed:
Use options selector instead of string for source ID
Fix test docstring

mj23000 committed:
Update options

andrewsayre commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -985,12 +989,24 @@ async def async_unjoin_player(self) -> None:
+        elif beolink_jid and source_id:
+            # Home Assistant does not support the "raw" Beolink Converter NL/ML source ids in strings.json
+            # So they have the be transformed to be compatible and now have to be capatilized in order to work
+            if BL_CONVERTER_PREFIX in source_id:
```
Two things:

1. Explicitly check for the prefix (vs. a substring):
```python
if source_id.startswith(BL_CONVERTER_PREFIX):
```

2. Please add a test to cover this unique functionality.

andrewsayre replied:
See comment below first. Looking to better understand why this is necessary. :)

mj23000 replied:
Yes, I'll add a test for that.

andrewsayre commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -29,6 +29,22 @@
+        "spotify": "Spotify",
+        "tidal": "Mozart Tidal",
+        "beoradio": "ASE Beoradio",
+        "bc_nlml_radio": "Beolink Converter NL/ML RADIO",
+        "bc_nlml_tp1": "Beolink Converter NL/ML TP1",
+        "bc_nlml_tp2": "Beolink Converter NL/ML TP2",
+        "bc_nlml_cd": "Beolink Converter NL/ML CD",
+        "bc_nlml_aux_a": "Beolink Converter NL/ML AUX_A",
+        "bc_nlml_ph": "Beolink Converter NL/ML PH"
```
I'm not following why the prefix (and associated removal logic) is necessary? Why wouldn't we want the user to be able to just indicate `radio` vs `bc_nlml_radio`?

mj23000 replied:
Only the Beolink Converter NL/ML sources have to be capitalized, so `radio` has to become `RADIO`. The prefix is simply To differentiate between sources that need to be transformed without for example defining a list of source_id's that should be transformed or a translation dict

andrewsayre replied:
Ah OK. thanks for clarifying. From a user perspective I think it's confusing and unnecessary. It would be better to use generic sources that would be similar across integrations. I'd rather see a translation dictionary or a list of values that need to be capitalized in this case. If you can influence the upstream library, that's logic that could also belong there instead of in HA.

mj23000 replied:
I disagree. This parameter is in an "advanced" section and these specific sources are for a more than ten year old product, which enables control of even older products, such as the Beosound 9000 which was released in 1996. I do not think most users will need this feature and would prefer not to complicate this further. The "weird" prefix will also only be exposed if users choose to use YAML configuration directly.

tjhorner replied:
From a developer's perspective, I agree with @andrewsayre — a translation dictionary is more explicit and easier to follow for future maintainers. The prefix-based rule seems too implicit and difficult to discover unless you are already aware of it.

From a user's perspective I am indifferent on the YAML key naming, but they should be internally consistent. What are the typically available sources when using `media_player.select_source`? I would expect those to be similar, so the `source_id` here should match whatever convention the media player's source IDs look like.

andrewsayre replied:
@mj23000 I hear what you're saying, however, let's solve the problem with consistency to other integrations. I think you'll find it's less complex (and less code) to check if the value is in a constant known list of values that need to be capitalized. It will also be more maintainable and less confusing to the user.

Perhaps something like this: 

```python
# Sources that need to be capitalized 
CONVERTER_SOURCES = [
    "radio",
    "tp1",
    "tp2",
    ...
]

...

# Beolink Converter NL/ML sources need to be capitalized.
if source_id.lower() in CONVERTER_SOURCES:
    source_id = source_id.upper()
```

I appreciate you changing the approach.

mj23000 replied:
Here is an overview of the raw source_ids available and their platforms:
- Mozart (This integration and consistent with `select_source`)
  - deezer
  - spotify
  - tidal
- ASE
  - deezer
  - spotify
  - beoradio
- Beolink Converter NL/ML
  - RADIO
  - TP1
  - TP2
  - CD
  - AUX_A
  - PH

Using a translation dict could look like (with the dict definition hidden away in const.py(?)):

```python
# Beolink Converter NL/ML source ids need to be in upper case
beolink_converter_map = {
    "radio": "RADIO",
    "tp1": "TP1",
    "tp2": "TP2",
    "cd": "CD",
    "aux_a": "AUX_A",
    "ph": "PH",
}
if source_id in beolink_converter_map:
    source_id = beolink_converter_map[source_id]
```

The problem with this is that looking at the YAML values, "radio" does not indicate that the source only works the Beolink Converter NL/ML.

So should all sources have some kind of prefix to indicate what platform they're compatible with? Then the translation map should be extended to all of the sources.

It feels like too much work when only a handful of source_ids need to be modified. 

Really the root problem is the hassfest rules for strings.json. I can understand why they don't allow uppercase keys, but that can add a layer of unnecessary work for the integration.

mj23000 replied:
OK. I've slept on it and I've added a translation dict, hardware platform prefix, service input validation, (hopefully) some better comments description and some more tests.

andrewsayre replied:
> So should all sources have some kind of prefix to indicate what platform they're compatible with? Then the translation map should be extended to all of the sources.

I don't think a prefix is necessary since the string labels and documentation will indicate which devices support it. However, since there's no overlap in sources types between devices in the final code (i.e. `deezer` only shows up once), what you've done is fine. Otherwise, if a source like `deezer` is supported across devices, then it would be better to not have prefixes and go with an approach similar to what I suggested above.

andrewsayre replied:
>  Otherwise, if a source like deezer is supported across devices, then it would be better to not have prefixes and go with an approach similar to what I suggested above.

Oh, I see the prefixes are actual permutations of devices platforms. **This is not the way.** 

Please keep the `select_source` values agnostic to the underlying platform type (without prefixes). The string descriptions and documentation is where to indicate what specific devices support it.

The only problem that needs to be solved is knowing when to capitalize certain input types (which is a device-specific detail) as the inputs are already device-type agnostic and can be passed to the library. As suggested above, a list of values that need to be capitalized solves this problem and doesn't leak implementation details.

mj23000 committed:
Use translation dict for source ids
Add input validation
Add tests for invalid sources
Improve source id description

andrewsayre commented on homeassistant/components/bang_olufsen/services.yaml:
```diff
@@ -48,6 +48,23 @@ beolink_join:
+            select:
+              translation_key: "source_ids"
+              options:
+                - ase_mozart_deezer
+                - ase_mozart_spotify
+                - mozart_tidal
+                - ase_beoradio
+                - beolink_converter_radio
+                - beolink_converter_tp1
+                - beolink_converter_tp2
+                - beolink_converter_cd
+                - beolink_converter_aux_a
+                - beolink_converter_ph
```
Please sort these for maintainability.

andrewsayre reviewed (changes requested):
Please keep the select_source values agnostic to the underlying platform type (without prefixes). The string descriptions and documentation is where to indicate what specific devices support it.

The only problem that needs to be solved is knowing when to capitalize certain input types (which is a device-specific detail) as the inputs are already device-type agnostic and can be passed to the library. As suggested above, a list of values that need to be capitalized solves this problem and doesn't leak implementation details.

mj23000 committed:
Use list instead of translation dict
Remove platform prefixes
Add test for Beolink Converter source

andrewsayre reviewed (approved):
Latest changes look good. PR is still marked draft, so won't merge until you indicate it's ready.

mj23000 committed:
Fix source_id naming and order

joostlek commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -61,6 +77,10 @@
+        },
+        "source_id": {
+          "name": "Source",
+          "description": "Specify which source to join, behavior varies between hardware platforms. Source names prefaced by a platform name can only be used when connecting to that platform. For example \"ASE Beoradio\" can only be used when joining an ASE device, while ”ASE / Mozart Deezer” can be used with ASE or Mozart devices. A defined Beolink JID is required."
```
Why does this sound so complicated? How is this displayed in the Bang & Olufsen app? Is this logical for the average user? Can we use validation to make sure this works well?

mj23000 replied:
The Beolink Converter NL/ML is a legacy product, designed to link the old B&O multiroom system from the 90’s to the current one. It is considered an advanced product, and the average user does not have one.
It does however enable current B&O products to link to legacy  sources, something that the B&O app also allows.
There is input validation through HA and if the source can't be joined, then the device will give an audible error indication.

andrewsayre reviewed (approved):
Thanks for answering our questions and making the revisions! Very much appreciate your contribution!
